feat(iota-indexer): add `reliability.md` (#6694)
iotaledger/iota: crates/iota-indexer/reliability.md

| ------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------- |

> | **1. Checkpoint Sync Lag**      | `latest_fullnode_checkpoint_sequence_number - max_committed_checkpoint_sequence_number`                                                                                                                                                                                                                                                                                                                                                                                                                              | - **Genesis**: Lag can be > 0 not more than 35 minutes <br><br>- **Post Genesis**: Lag can be > 0 not more than 5 minutes              | Should be a Time series panel |

Reviewer: If I get this right, the metric that should trigger an action is the time that we observe any lag between the `latest_fullnode` and the `max_commited` checkpoint.

If that is the case, I think the `Metrics` column should explicitly mention that. And the threshold then should adapt accordingly. Something like 
```suggestion
| **1. Checkpoint Sync Lag**      | Time in `min` where `lag > 0`. `lag` is defined as `latest_fullnode_checkpoint_sequence_number - max_committed_checkpoint_sequence_number`                                                                                                                                                                                                                                                                                                                                                                                                                              | - **Genesis**: `> 35 min` <br><br>- **Post Genesis**: `> 5 min`              | Should be a Time series panel |
```

> | **1. Checkpoint Sync Lag**      | `latest_fullnode_checkpoint_sequence_number - max_committed_checkpoint_sequence_number`                                                                                                                                                                                                                                                                                                                                                                                                                              | `t` represent the time in minutes.<br>`L(t)` represent the lag size at time `t`.<br><br>- **Genesis**: 0 ≤ t ≤ 35, such that L(t) > 0 <br><br>- **Post Genesis**: 0 ≤ t ≤ 5, such that L(t) > 0                 | Should be a Time series panel |

Reviewer 2: I'm not sure how it should be interpreted:

```**Genesis**: 0 ≤ t ≤ 35, such that L(t) > 0```

Could you explain in other words?

Author: - The lag `L(t)` should return to zero within 35 minutes from when the indexer started processing the genesis checkpoint.
- In other words, the system is allowed up to 35 minutes to complete processing the genesis checkpoint data, the the lag between full node and indexer can be > 0 as long it does not exceed 35 min.

for example:
`L(10) > 0` can be interpreted as: The lag size between full node and indexer is greater than 0 for at least 10 minutes

Let me know if I should add the example in the **Notes** column.

Reviewer 2: Maybe we could formulate the Thresholds using natural language, not mathematical one? I feel it would be hard to interpret it like this.

Author: fixed: https://github.com/iotaledger/iota/pull/6694/commits/79e3b48eb5235cda2d0828cf70b1d3966be03e31

Reviewer 2: I think it's more clear now, thanks.

> | **2. Data Ingestion Latencies** | `checkpoint_db_commit_latency`<br> `checkpoint_db_commit_latency_transactions`<br> `checkpoint_db_commit_latency_tx_insertion_order`<br> `checkpoint_db_commit_latency_objects`<br> `checkpoint_db_commit_latency_objects_snapshot`<br> `checkpoint_db_commit_latency_objects_history`<br> `checkpoint_db_commit_latency_events`<br> `checkpoint_db_commit_latency_event_indices`<br> `checkpoint_db_commit_latency_packages`<br> `checkpoint_db_commit_latency_tx_indices`<br> `checkpoint_db_commit_latency_epoch` | - **Genesis**: ≤ 15 minutes.<br><br>- **Post Genesis**: ≤ 15 seconds.                                                                  |                               |

Reviewer: Wouldn't it make sense to have a separate threshold for each latency?

What do these represent? The commit for a single transaction or a  batch of transactions? Is `15 sec` acceptable for all?

Author: fixed: https://github.com/iotaledger/iota/pull/6694/commits/8302fab3e7175af425c09d9564c90b04b405723f

> | **3. DB Connection Pool Size**  | `db_conn_pool_size`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  | - **Genesis sync**: Pool size can be 0 not more than 10 seconds <br><br>- **Post-genesis**: Pool size can be 0 not more than 5 seconds | Should be a Time series panel |

Reviewer: Same as in the `Checkpoint Sync Lag`. The metric is the time where pool-size = 0. So I suggest this is adapted accordingly.

Author: fixed: https://github.com/iotaledger/iota/pull/6694/commits/8302fab3e7175af425c09d9564c90b04b405723f

### Metrics

The following indexer metrics will be used for quantitative requirements.

| **Metric**                                               | **Description**                                                                                                     |
| -------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- |
| `latest_fullnode_checkpoint_sequence_number`             | Latest checkpoint from full node ready to be synced by the indexer.                                                 |
| `max_committed_checkpoint_sequence_number`               | The highest committed checkpoint by the indexer.                                                                    |
| `checkpoint_db_commit_latency`                           | Total latency to commit a chunk of checkpoints (including all transactions, objects, events, etc.) to the database. |
| `checkpoint_db_commit_latency_transactions_chunks`       | Latency to commit a single chunk of checkpoint transactions to the database.                                        |

> | `checkpoint_db_commit_latency_tx_insertion_order_chunks` | Latency to commit as single chunk of data representing the transaction insertion order to the database.             |

Reviewer: ```suggestion
| `checkpoint_db_commit_latency_tx_insertion_order_chunks` | Latency to commit a single chunk of data representing the transaction insertion order to the database.             |
```

| `checkpoint_db_commit_latency_objects_chunks`            | Latency to commit a single chunk of data representing the objects to the database.                                  |
| `checkpoint_db_commit_latency_objects_history_chunks`    | Latency to commit a single chunk of data representing the object history to the database.                           |
| `checkpoint_db_commit_latency_objects_snapshot_chunks`   | Latency to commit a single chunk of data representing the object snapshot to the database.                          |
| `checkpoint_db_commit_latency_objects_snapshot`          | Total latency to commit a batch of checkpoints object snapshots to the database.                                    |
| `checkpoint_db_commit_latency_epoch`                     | Latency to commit the epoch data into database.                                                                     |
| `req_latency_by_route`                                   | Latency of indexer JSON RPC API endpoints. Measures the time from request receipt to response delivery.             |

> [!Note]
>

> > - Large datasets (e.g., transactions, objects) are split into smaller **chunks** to avoid overwhelming the database with bulk writes.

Reviewer: What is the default chunk size?

Author: fixed: https://github.com/iotaledger/iota/pull/6694/commits/96fa180bfbd0c6d8db38c51a06afacd9b43ac2b9

- The primary pipeline uses by default chunks of **100** elements, it can be modified by this env var `PG_COMMIT_PARALLEL_CHUNK_SIZE`.
- The object snapshot pipeline uses by default chunks of **500** elements, it can be modified by this env var `PG_COMMIT_OBJECTS_PARALLEL_CHUNK_SIZE`.

> - Metrics ending with `_chunks` track latency **per individual chunk**, not the entire dataset of the checkpoint being committed.
> - Example: `checkpoint_db_commit_latency_transactions_chunks` measures the time to commit **one transaction chunk**, not all transactions in a checkpoint.

### Quantitative Requirements:

> | **Requirement****                                                      | **Metrics**                                                                                                                                        | **Permissible Values**                                               |

Reviewer: ```suggestion
| **Requirement**                                                      | **Metrics**                                                                                                                                        | **Permissible Values**                                               |
```
